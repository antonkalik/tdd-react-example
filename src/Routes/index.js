import { Routes as ReactRoutes, Route, Outlet } from "react-router-dom";
import { HomeView } from "src/views/HomeView";
import { CardsView } from "src/views/CardsView";
import { CardView } from "src/views/CardView";
import { AboutView } from "src/views/AboutView";
import { MainLayout } from "src/components/MainLayout";

export const Routes = () => {
  return (
    <ReactRoutes>
      <Route path="/" element={<MainLayout />}>
        <Route index element={<HomeView />} />
        <Route path="/cards" element={<Outlet />}>
          <Route index element={<CardsView />} />
          <Route path=":id" element={<CardView />} />
        </Route>
        <Route path="/about" element={<AboutView />} />
      </Route>
      <Route path="*" element={<h1>404 - Not Found</h1>} />
    </ReactRoutes>
  );
};
