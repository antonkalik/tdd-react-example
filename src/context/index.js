import React, { useReducer } from "react";

export const initialState = { cards: [] };

export function reducer(state, action) {
  switch (action.type) {
    case "ADD_CARD":
      return {
        cards: action.payload.title
          ? [
              ...state.cards,
              {
                ...action.payload,
                id: state.cards.length + 1,
              },
            ]
          : state.cards,
      };
    case "REMOVE_CARD":
      return {
        cards: state.cards.filter((card) => card.id !== action.payload.id),
      };
    case "RESET":
      return initialState;
    default:
      return state || initialState;
  }
}

export const AppContext = React.createContext();

const StoreProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const addCard = (title) => {
    dispatch({ type: "ADD_CARD", payload: { title } });
  };

  const removeCard = (id) => {
    dispatch({ type: "REMOVE_CARD", payload: { id } });
  };

  const reset = () => {
    dispatch({ type: "RESET" });
  };

  const value = {
    state,
    actions: {
      addCard,
      removeCard,
      reset,
    },
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
};

export default StoreProvider;
