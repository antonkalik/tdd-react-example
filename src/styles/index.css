html,
body {
  background-color: #222;
  padding: 0;
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen,
    Ubuntu, Cantarell, Fira Sans, Droid Sans, Helvetica Neue, sans-serif;
}

.app {
  padding: 20px;
}

a,
p,
h1,
h2,
h3,
h4,
h5,
h6,
div {
  color: #fff;
}
