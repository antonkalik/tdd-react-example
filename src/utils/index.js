export const initCards = Array.from({ length: 5 }, (_, i) => ({
  title: `Test Card ${i + 1}`,
  id: i + 1,
}));
