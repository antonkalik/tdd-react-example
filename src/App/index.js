import React from "react";
import { BrowserRouter } from "react-router-dom";
import { Routes } from "src/Routes";
import { StyledApp } from "./style";
import StoreProvider from "../context";

export function App() {
  return (
    <StyledApp data-testid="app">
      <StoreProvider>
        <BrowserRouter>
          <Routes />
        </BrowserRouter>
      </StoreProvider>
    </StyledApp>
  );
}
