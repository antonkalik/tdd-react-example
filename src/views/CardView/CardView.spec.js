import { render, screen } from "@testing-library/react";
import { AppContext } from "src/context";
import { CardView } from "./index";

jest.mock("react-router-dom", () => ({
  useParams: () => ({
    id: 9,
  }),
}));

describe("<CardView />", () => {
  it("should render the card view", () => {
    const { getByTestId } = render(
      <AppContext.Provider
        value={{
          state: {
            cards: [
              {
                id: 9,
                title: "Card 9",
              },
            ],
          },
        }}
      >
        <CardView />
      </AppContext.Provider>
    );
    expect(getByTestId("card-view")).toMatchSnapshot();
    expect(getByTestId("card-view")).toHaveTextContent("Card 9");
  });

  it("should show not found", () => {
    const { getByTestId } = render(
      <AppContext.Provider value={{ state: { cards: [] } }}>
        <CardView />
      </AppContext.Provider>
    );
    expect(getByTestId("card-view")).toMatchSnapshot();
    expect(getByTestId("card-view")).toHaveTextContent("Card not found");
  });
});
