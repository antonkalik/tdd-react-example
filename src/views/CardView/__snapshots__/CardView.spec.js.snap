// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardView /> should render the card view 1`] = `
<div
  data-testid="card-view"
>
  <p
    data-testid="card-view-id"
  >
    9
  </p>
  <h1>
    Card 9
  </h1>
</div>
`;

exports[`<CardView /> should show not found 1`] = `
<div
  data-testid="card-view"
>
  <h1>
    Card not found
  </h1>
</div>
`;
